NotebookAnalyzers/utils.py

```diff
@@ -281,3 +273,114 @@ def classes_number(code):
     tree = ast.parse(code)
     class_def_num = sum(isinstance(exp, ast.ClassDef) for exp in tree.body)
     return class_def_num
+
+
+def _non_executed_cells_count(cell_list):
+    """The function takes a list of cells and returns the number of non-executed cells
+
+        Args:
+            cell_list(list): list of dictionary objects representing the notebook cells
+        Returns:
+            non_exec_cells: number of non-executed cells in the list
+        
+        The function is used in:
+        - count_non-executed_cells(nb_dict)
+        - count_bottom_non-executed_cells(nb_dict, bottom_size=4)          
+    """
+    non_exec_cells = 0
+    for cell in cell_list:
+        if cell["cell_type"] == 'code':
+            if cell['execution_count'] is None and cell['source'] != []:
+                non_exec_cells = non_exec_cells + 1  # This is a not executed Python Cell containing actual code
+    return non_exec_cells
+
+
+def _empty_cells_count(cell_list):
+    """The function takes a list of cells and returns the number of empty cells
+
+        Args:
+            cell_list(list): list of dictionary objects representing the notebook cells
+        Returns:
+            empty_cells: number of empty cells in the list
+
+        The function is used in:
+        - count_empty_cells(nb_dict)
+        - count_bottom_empty_cells(nb_dict, bottom_size=4)          
+    """
+    empty_cell = 0
+    for cell in cell_list:
+        if cell["cell_type"] == 'code':
+            if cell['execution_count'] is None and cell['source'] == []:
+                empty_cell = empty_cell + 1  # This is an empty Python Cell
+    return empty_cell
+
+
+def count_bottom_non_executed_cells(nb_dict, bottom_size=4):
+    """
+        Number of non-executed cells between the last bottom-size-cells of the notebook
+
+        Args:
+            nb_dict(dict): python dictionary object representing the jupyter notebook
+            bottom_size(int): number of cells starting from the bottom of the dictionary
+        Returns:
+            _non_executed_cells_count(cell_list): number of non-executed cells in the bottom-size last section of the notebook
```

Comment:
Nella documentazione di questa funzione andrebbe specificata la precondizione:

```python
if bottom_size < (len(nb_dict["cells"]) / 3)
```

(e ciò che succede nei casi in cui la stessa non sia soddisfatta).